### DISCUSSION TITLE
Issue Updating Firmware with cfclient - Stuck in Infinite Loop AIdeck

### DISCUSSION THREAD

**Discussion by gelanchez** [bitcraze/discussions#1573] [Q&A]
Hi,

I'm experiencing an issue when updating the firmware using the cfclient, following these instructions: [Getting Started with AIdeck](https://www.bitcraze.io/documentation/tutorials/getting-started-with-aideck/#update-crazyflie-and-aideck-firmware).

The console log shows an infinite loop of messages saying that it wasn't possible to open the camera on the GAP8. This message comes from the firmware I tried to flash (based on the example usage of the camera and WiFi transmission), but the process got stuck at 99%. I’ve noticed it usually gets stuck at either 99% or 3% when flashing over WiFi, so I always reflash the GAP8 bootloader ([instructions here](https://www.bitcraze.io/documentation/tutorials/getting-started-with-aideck/#gap8-bootloader)) and then reflash the firmware.

However, this time I’m stuck in an infinite loop and can't resolve the issue. What else can I try?

Many thanks,
Jose

**Comment by gelanchez** [comment]
Hi,

I found a way to escape the infinite loop by flashing a working GAP8 firmware via JTAG, which allows me to reflash the Crazyflie firmware using cfclient. However, I think the issue starts after flashing the GAP8 bootloader ([instructions here](https://www.bitcraze.io/documentation/tutorials/getting-started-with-aideck/#gap8-bootloader)), as I can’t get past 99%. Oddly enough, I was able to complete this procedure before, but it seems the bootloader might have become damaged.

The problem persists: when flashing over WiFi, it consistently gets stuck at 99%. To avoid the infinite loop, I flash a GAP8 firmware via JTAG that doesn't start the camera, but then cfclient firmware doesn't progress when trying to update the AIdeck firmware (it can't find the deck). I’ve tried several combinations: GAP8 bootloader via JTAG and Docker, CF firmware with cfclient, cold boot and GAP8 firmware over JTAG, and the furthest I’ve gotten is being stuck at 99% when using cfloader flashing.

Could you suggest anything else I can try? I’ve hit a dead end.

Thanks!

**Reply by evoggy** [reply]
Hi!

Thanks for all the information! Just to double check, are you using the latest version of libraries and firmware from git? Including the latest GAP8 bootloader and the latest STM32 firmware?

**Reply by gelanchez** [reply]
Hi, @evoggy. Yes, I'm using the latest version of the GAP8 bootloader and the firmware I flash with the cfclient is also the latest one.

**Reply by gemenerik** [reply]
> I think the issue starts after flashing the GAP8 bootloader ([instructions here](https://www.bitcraze.io/documentation/tutorials/getting-started-with-aideck/#gap8-bootloader)), as I can’t get past 99%

Do I understand correctly that you can flash the GAP8 bootloader over JTAG successfully? And what you are trying to do is flashing a release through the client-but this fails to flash the AI-deck because the deck is not detected? And it is OTA flashing of a GAP8 example/app that gets stuck at 99%?

If you connect to your Crazyflie and in the client check the console, do you see printouts regarding the DECK_CORE detecting the AI-deck?
Does the client with which you run the client print out anything regarding the flashing attempt failling? Can you share that here?

**Reply by gelanchez** [reply]
Hi, @gemenerik,

I think I might have corrupted the code / memory of the GAP8 somehow as the main problem I have is that the client can't flash anything there. When I try, the output loops infinitely after a failure:

```
SYS: ----------------------------
SYS: Crazyflie 2.1 is up and running!
SYS: Build 0:80ee72ad480a (2024.10) CLEAN
SYS: I am 0x38313238343947190027003B and I have 1024KB of flash!
CFGBLK: v1, verification [OK]
DECK_CORE: 1 deck(s) found
DECK_CORE: Calling INIT on driver bcAI for deck 0
CPX-EXT-ROUTER: Waiting for CPX External router initialization
CPX-EXT-ROUTER: CPX External router initialized, CPX_VERSION: 0
AIDECK: Not setting up WiFi
IMU: BMI088: Using I2C interface.
IMU: BMI088 Gyro connection [OK].
IMU: BMI088 Accel connection [OK]
IMU: BMP388 I2C connection [OK]
ESTIMATOR: Using Complementary (1) estimator
CONTROLLER: Using PID (1) controller
MTR-DRV: Using brushed motor driver
SYS: About to run tests in system.c.
SYS: NRF51 version: 2024.10 (CF21)
EEPROM: I2C connection [OK].
STORAGE: Storage check [OK].
IMU: BMI088 gyro self-test [OK]
DECK_CORE: Deck 0 test [OK].
SYS: Self test passed!
STAB: Wait for sensor calibration...
SYS: Free heap: 12032 bytes
CPX: ESP32: I (819) SYS: Initialized
CPX: ESP32: I (819) DISCOVERY: Hostname is aideck-CC46B8
CPX: ESP32: I (819) DISCOVERY: MAC (STA) is 3c:71:bf:cc:46:b8
CPX: GAP8: 
CPX: GAP8: Setting up WiFi AP
CPX: GAP8: Task CAMERA_TASK created
CPX: ESP32: I (1339) WIFI: wifi_init_softap finished
CPX: GAP8: Opening Himax camera
CPX: GAP8: Failed to open camera
CPX: GAP8: Failed to open and configure camera
CPX: GAP8: Task CAMERA_TASK created
CPX: GAP8: Opening Himax camera
CPX: GAP8: Failed to open camera
CPX: GAP8: Failed to open and configure camera
CPX: GAP8: Task CAMERA_TASK created
CPX: GAP8: Opening Himax camera
CPX: GAP8: Failed to open camera
...
```
Flashing over radio with the cfloader doesn't work at all. I can flash the GAP8 via the JTAG a modified version of my software in which I don't enable the camera. After this, the cfloader tool works, but only until 99% with the wifi-img-streamer example, so I have followed [this](https://www.bitcraze.io/documentation/tutorials/getting-started-with-aideck/#gap8-bootloader). (I thought this was necessary only once) and flashed the GAP8 bootloader, but then I go back to step 0, with that infinite loop when trying to flash the firmware. If I skip the step of flashing the GAP8 bootloader and go straight to flashing the firmware, that now works, but the cfloader still stays at 99%.

Sorry for my long explanation, I hope I've made the problem understandable :) What else can I try? And many thanks for your help!

**Reply by gemenerik** [reply]
So you are still getting stuck at 99% while e.g. [flashing the wifi example](https://www.bitcraze.io/documentation/tutorials/getting-started-with-aideck/#flash-wifi-example), even after reflashing the bootloader? Can you try a [firmware update through the client](https://www.bitcraze.io/documentation/repository/crazyflie-clients-python/master/userguides/userguide_client/#firmware-upgrade) with the AI-deck attached? This will flash the ESP32, too. Then retry flashing the wifi example for example.

**Reply by gelanchez** [reply]
Hi, @gemenerik. That doesn't work, it's part of the procedure I followed before. After updating the firmware with the AI-deck attached through the client, flashing the wifi example still gets stuck at 99%.

**Reply by Bananazooka** [reply]
Hi, 
Have you tried [this section ](https://www.bitcraze.io/documentation/repository/aideck-gap8-examples/master/infrastructure/flashing/)of the tutorial? This will flash the firmware, not the bootloader.

**Reply by gelanchez** [reply]
Thanks, @Bananazooka, I've also tried those. After many attempts, I've narrowed down the steps that work and don't:

1. Flashing  _wifi-img-streamer_ over JTAG works and removes the infinite loop showed in the cfclient.
2. Flashing anything over radio always hangs up, usually at 99%. When that happens, following the [tutorial](https://www.bitcraze.io/documentation/tutorials/getting-started-with-aideck/#gap8-bootloader) on the Gap8 bootloader makes the infinite loop appear again when trying to flash the firmware with the cfclient ([tutorial](https://www.bitcraze.io/documentation/tutorials/getting-started-with-aideck/#update-crazyflie-and-aideck-firmware)) (oddly enough the old code messages show up !!). In that case, I need to do step 1 again. If I just try flashing the firmware without touching the bootloader, that works, but the radio flashing problem doesn't get solved.

Is there anything else I can try to fix this flashing over radio problem?

Many thanks for all your help!

**Reply by gemenerik** [reply]
Hi José. I'm sorry to hear it's still not working. Please send us an e-mail at support@bitcraze.io and link this thread, we'll send you a replacement.
